File: 1858_longest_word_with_all_prefixes/Solution.swift
// LeetCode 1858 - Longest Word With All Prefixes
// https://leetcode.com/problems/longest-word-with-all-prefixes/

class Solution {
    func longestWord(_ words: [String]) -> String {
        let wordSet = Set(words)
        var best = ""

        for word in words {
            var prefix = word
            var valid = true
            while !prefix.isEmpty {
                if !wordSet.contains(prefix) {
                    valid = false
                    break
                }
                prefix.removeLast()
            }

            if valid && (word.count > best.count || (word.count == best.count && word < best)) {
                best = word
            }
        }

        return best
    }
}

File: 1858_longest_word_with_all_prefixes/solution.ts
// LeetCode 1858 - Longest Word With All Prefixes
// https://leetcode.com/problems/longest-word-with-all-prefixes/

function longestWord(words: string[]): string {
    const wordSet = new Set(words);
    let best = "";
    for (const word of words) {
        let prefix = word;
        let valid = true;
        while (prefix) {
            if (!wordSet.has(prefix)) {
                valid = false;
                break;
            }
            prefix = prefix.slice(0, -1);
        }
        if (valid && (word.length > best.length || (word.length === best.length && word < best))) {
            best = word;
        }
    }
    return best;
}
